homeworks/solutions/torantie/3-smooth/src/halfedge.js

```diff
@@ -0,0 +1,632 @@
+      this.vertsOrig.push(vertexCopy)
+    }
+  }
+
+
+  laplaceMatrix(weightType) {
+    // TODO: construct the Laplace matrix.
+    const numberOfVertices = this.vertices.length
+    let weightTriplet = new Triplet(numberOfVertices, numberOfVertices)
+
+    for (const vert of this.vertices) {
+      const i = vert.idx
+      let sum = 1e-8 // Tikhonov regularization to get strict positive definite
+
+      vert.forEachHalfEdge(h => {
+        let w = 0
+
+        switch (weightType) {
+          case 'uniform':
+            w = 1
+            break
+          case 'cotan':
+            w = (h.cotan() + h.twin.cotan())/2
+        }
+
+        sum += w
+        //console.log("Value: "+w+" Position: ("+ i+","+h.twin.vertex.idx +")")
+        weightTriplet.addEntry(w, i, h.twin.vertex.idx)
+      })
+      weightTriplet.addEntry(-sum, i, i)
+      //console.log("Value: "+-sum+" Position: ("+ i+","+i +")")
+    }
+
+    let weightMatrix = SparseMatrix.fromTriplet(weightTriplet)
+
+    return weightMatrix
+  }
+
+  massMatrix(weightType){
+    const numberOfVertices = this.vertices.length
+    let massTriplet = new Triplet(numberOfVertices, numberOfVertices)
+
+    for (const vert of this.vertices) {
+      const i = vert.idx
+
+      switch (weightType) {
+        case 'uniform':
+          let neighbours = 0
+          vert.forEachHalfEdge(() => {
+            neighbours++
+          })
+
+          massTriplet.addEntry(neighbours,i,i)
+          break
+        case 'cotan':
+
+          let area = vert.calculateVoronoiArea()
+          // area alone is very small and destroys the smoothing
+          massTriplet.addEntry(1+area,i,i)
+          break
+      }
+    }
+
+    let massMatrix = SparseMatrix.fromTriplet(massTriplet)
+
+    return massMatrix
+  }
+
+
+  /**
+   * smooth performs the Laplacian smoothing algorithm.
+   * @param {string} weightType indicates the type of the weight for
+   * constructing the Laplace matrix. Possible value could be: 'uniform',
+   * 'cotan'.
+   * @param {Number} timeStep the time step in Laplacian Smoothing algorithm
+   * @param {Number} smoothStep the smooth step in Laplacian Smoothing algorithm
+   */
+  smooth(weightType, timeStep, smoothStep) {
+    //reset vertices
+    for (let i = 0; i < this.vertsOrig.length; i++) {
+      this.vertices[i].position.x = this.vertsOrig[i].position.x
+      this.vertices[i].position.y = this.vertsOrig[i].position.y
+      this.vertices[i].position.z = this.vertsOrig[i].position.z
+    }
+
+    for (let j = 0; j< smoothStep; j++) {
+      // TODO: implmeent the Laplacian smoothing algorithm.
+      //
+      try {
+        // Hint:
+        //
+        //   1. Construct the Laplace matrix `L` for the given `weightType`
+        let W = this.laplaceMatrix(weightType)
+        let M = this.massMatrix(weightType)
+        //   2. Solve linear equation: (I-tλL) f' = f using a Cholesky solver.
+        let f = M.minus(W.timesReal(timeStep))
```

Comment:
As an additional note (for your further reading):

The smooth step and time step can be improved. In each smooth step, the Laplace matrix is recomputed.
This is essential because the positions are updated after each step, and the matrix should also be updated/recomputed.

There are research papers actually investigate what could happen if the Laplace matrix is only computed once (always use the original Laplace). See, for instance:

Kazhdan, Michael, Jake Solomon, and Mirela Ben Chen.
"Can mean curvature flow be modified to be non-singular?."
Computer Graphics Forum. 2012. https://arxiv.org/pdf/1203.6819.pdf